src/ccs/lcccs.rs

```diff
@@ -23,67 +23,83 @@ use core::{cmp::max, marker::PhantomData};
+  pub(crate) r_x: Vec<G::Scalar>,
+  pub(crate) z: Vec<G::Scalar>,
 }
 
 impl<G: Group> LCCCS<G> {
-  // XXX: Double check that this is indeed correct.
-  /// Samples public parameters for the specified number of constraints and variables in an CCS
-  pub fn commitment_key(&self) -> CommitmentKey<G> {
-    let total_nz = self.ccs.M.iter().fold(0, |acc, m| acc + m.coeffs().len());
-
-    G::CE::setup(b"ck", max(max(self.ccs.m, self.ccs.t), total_nz))
+  /// Generates a new LCCCS instance from a given randomness, CommitmentKey & witness input vector.
+  /// This should only be used to probably test or setup the initial NIMFS instance.
+  pub(crate) fn new<R: RngCore>(
+    ccs: &CCS<G>,
+    ccs_m_mle: &[MultilinearPolynomial<G::Scalar>],
+    ck: &CommitmentKey<G>,
+    z: Vec<G::Scalar>,
+    mut rng: &mut R,
+  ) -> Self {
+    // XXX: API doesn't offer a way to handle this??
+    let _r_w = G::Scalar::random(&mut rng);
+    let w_comm = <<G as Group>::CE as CommitmentEngineTrait<G>>::commit(ck, &z[(1 + ccs.l)..]);
+
+    // Evaluation points for `v`
+    let r_x: Vec<G::Scalar> = (0..ccs.s).map(|_| G::Scalar::random(&mut rng)).collect();
+    let v = ccs.compute_v_j(&z, &r_x, ccs_m_mle);
+
+    Self { w_comm, v, r_x, z }
   }
 
   /// Checks if the CCS instance is satisfiable given a witness and its shape
-  pub fn is_sat(&self, ck: &CommitmentKey<G>, W: &CCSWitness<G>) -> Result<(), NovaError> {
+  pub fn is_sat(
+    &self,
+    ccs: &CCS<G>,
+    ccs_m_mle: &[MultilinearPolynomial<G::Scalar>],
+    ck: &CommitmentKey<G>,
+  ) -> Result<(), NovaError> {
+    dbg!(self.z.clone());
+    let w = &self.z[(1 + ccs.l)..];
     // check that C is the commitment of w. Notice that this is not verifying a Pedersen
     // opening, but checking that the Commmitment comes from committing to the witness.
-    let comm_eq = self.C == CE::<G>::commit(ck, &W.w);
+    let comm_eq = self.w_comm == CE::<G>::commit(ck, w);
 
-    // check CCS relation
-    let z: Vec<G::Scalar> = [vec![self.u], self.x.clone(), W.w.to_vec()].concat();
-    let computed_v =
-      compute_all_sum_Mz_evals::<G>(&self.matrix_mles, &z, &self.r_x, self.ccs.s_prime);
+    let computed_v = compute_all_sum_Mz_evals::<G>(ccs_m_mle, &self.z, &self.r_x, ccs.s_prime);
+    dbg!(self.v.clone());
+    dbg!(computed_v.clone());
     let vs_eq = computed_v == self.v;
 
-    dbg!(vs_eq);
-    dbg!(comm_eq);
-
     if vs_eq && comm_eq {
       Ok(())
     } else {
       Err(NovaError::UnSat)
     }
   }
 
-  /// Compute all L_j(x) polynomials
-  // Can we recieve the MLE of z directy?
-  pub fn compute_Ls(&self, z: &Vec<G::Scalar>) -> Vec<VirtualPolynomial<G::Scalar>> {
-    let z_mle = dense_vec_to_mle(self.ccs.s_prime, z);
-
-    let mut vec_L_j_x = Vec::with_capacity(self.ccs.t);
-    for M_j in self.matrix_mles.iter() {
+  /// Compute all L_j(x) polynomials.
+  pub fn compute_Ls(
+    &self,
+    ccs: &CCS<G>,
+    ccs_m_mle: &[MultilinearPolynomial<G::Scalar>],
+    ck: &CommitmentKey<G>,
```

Comment:
Looks like `ck` is not used in the method although is passed as an input parameter, is this because of some reason?